### ISSUE TITLE
failure in ERS 3 month nocomp test

### ISSUE THREAD

**Issue by mvertens** [NGEET/fates#1051]
I am trying to create longer runs using the latest CTSM in NorESM configurations (with the latest CMEPS, CDEPS, etc). The following 2 tests fail restarts: 
ERS_Lm3.f45_f45_mg37.2000_DATM%GSWP3v1_CLM51%FATES-NOCOMP_SICE_SOCN_SROF_SGLC_SWAV_SESP.betzy_intel.clm-FatesColdNoCompNoFire

ERS_Ld90.f45_f45_mg37.2000_DATM%GSWP3v1_CLM51%FATES-NOCOMP_SICE_SOCN_SROF_SGLC_SWAV_SESP.betzy_intel.clm-FatesColdNoCompNoFire

However what is interesting is that this test passes restart:
ERS_Ld90.f45_f45_mg37.2000_DATM%GSWP3v1_CLM51%FATES-NOCOMP_SICE_SOCN_SROF_SGLC_SWAV_SESP.betzy_intel.clm-FatesColdNoCompNoFire.20230707_114133_1m5tec/

Since the restart test fails - we cannot really start an longer runs until we resolve this problem.

**Comment by mvertens** [comment]
@rosiealice @glemieux - I was not sure where to post this. The code is very close to ctsm5.1.dev129.

**Comment by rosiealice** [comment]
I guess a higher level question is whether we should push this NOCOMP compset back into CTSM to streamline the testing, given it (nocomp) is a pretty core use-case of FATES at the moment? Should we discuss this on a FATES SE call one day (those are at 9pm on Mondays in Norway Mariana)

I don't know what the difference is between these two test variants...

**Comment by glemieux** [comment]
Hi @mvertens, I'll try replicating this test.  As @rosiealice asked, what's the difference between the two `ERS_Ld90` tests?  Were they using different fates tags?

**Comment by mvertens** [comment]
@glemieux - thanks so much for your quick response to this.
Sorry I had wrong test written down incorrectly.
It should be 
ERS_Ld120.f45_f45_mg37.2000_DATM%GSWP3v1_CLM51%FATES-NOCOMP_SICE_SOCN_SROF_SGLC_SWAV_SESP.betzy_intel.clm-FatesColdNoCompNoFire
that fails and the following passes.
ERS_Ld90.f45_f45_mg37.2000_DATM%GSWP3v1_CLM51%FATES-NOCOMP_SICE_SOCN_SROF_SGLC_SWAV_SESP.betzy_intel.clm-FatesColdNoCompNoFire
I also tried ERS_Ld30 and ERS_Ld60 and those passed as well.

**Comment by glemieux** [comment]
I can confirm that I'm seeing similar COMPARE_base_rest failures using the following test case setup:

`ERS_Lm3.f45_f45_mg37.2000_DATM%GSWP3v1_CLM51%FATES_SICE_SOCN_SROF_SGLC_SWAV.cheyenne_intel.clm-FatesColdNoComp`

Here's what I checked varying the durations:
- `Lm3`: fails
- `Ld90`: passes
- `Ld120`: fails
- `Lm6`: fails

Here are other checks I did using `Lm3`:
- `CRUv7` versus `GSWP3v1`: both fail
- `FatesCold` versus `FatesColdNoComp`: `FatesCold` passes
- `ctsm5.1dev126` versus `ctsm5.1.dev129`: pending

**Comment by mvertens** [comment]
@glemieux - thanks for confirming this. The one way I have always used to iron out restart problems is to get close to where the problem is and write history files every time step - for an initial run and a restart run. I can try to narrow this down some more. Clearly stopping somewhere between day 90 and day 120 is causing the problem. So say we know that stopping at day 100 and restarting will not work. Lets assume that just looking at the day boundary is good enough.  Then one way to debug this is the following - assume that coupler history files and clm history files are written every time step.
1) Do a run until day 99 - and write write a restart at the end 
2) Restart from day 99 and run for 50 time (assuming half hour time steps)  and write a restart at day 100
3) Save the history files 
3) Now restart again at and run for 3 time steps. Compare the history files for time steps 48-50 and see where the differences start. 

I've usually been successful using this approach - and just did that to find a problem for a restart problem using a compset with ocean, ice, wave and datm. Anyway - I'm happy to help as well - and am also happy to clarify things if the explanation above has been confusing.
4) compare

**Comment by rgknox** [comment]
@mvertens do you know what grid-cell is showing the diff? 

Also, just  to confirm, I see you have a FatesCold test that passes with LM3.  Have you found any non-Nocomp tests failing, or are failures just nocomp configurations?

**Comment by mvertens** [comment]
@rgknox - yes that is correct. The only failures I have found are in the one nocomp configuration. I have only added 4 regressions tests at this point to the NorESM CTSM fork - as a way of moving forwards to run this with CAM60, CICE and BLOM.
ERS_D_Ld15.f45_f45_mg37.I2000Clm50FatesRs.betzy_intel.clm-FatesColdTreeDamage (Overall: PASS) details:
ERS_D_Ld3_PS.f09_g17.I2000Clm50FatesRs.betzy_intel.clm-FatesCold (Overall: PASS) details:
ERS_Lm13.f45_f45_mg37.I2000Clm50Fates.betzy_intel.clm-FatesColdNoComp (Overall: FAIL) details:
  FAIL ERS_Lm13.f45_f45_mg37.I2000Clm50Fates.betzy_intel.clm-FatesColdNoComp COMPARE_base_rest
SMS.f45_f45_mg37.I2000Clm51FatesSpRsGs.betzy_intel.clm-FatesColdSatPhen (Overall: PASS) details:

**Comment by glemieux** [comment]
A bunch of us met today to brainstorm next steps on how to approach this issue.  I reported that yesterday I ran a few ERS FatesColdNoComp test mods at different durations and narrowed down the restart to some time between 52 and 58 days (Ld100 and Ld110 respectively) and is showing up first in southern latitudes (New Zealand and south Argentina).  The issue presents itself right at the start up time.  There are some additional nearby gridcells that start showing issues towards the end of the run suggesting that this issue might be more wide spread in the boreal region, but just swamped out because of the plotting range.  The latest test runs are with the latest ctsm and fates tags.

Some additional notes from the meeting:
- Could we be missing a variable in the restart file?  Is the restart using the cold start value that just happens to be extremely small?
- Given the evidence that this is happening in the southern boreal region, could this have something to do with deciduous phenology?
- If the restart issue appears using a fully spun up case, we probably should accelerate addressing https://github.com/ESCOMP/CTSM/issues/1277

The following actions are going to be run:
1. @mvertens will run a restart tests with a spin up data set from @rosiealice
2. @rgknox will investigate running with evergreens only

We've agreed to meet in two weeks time to discuss progress.

**Comment by mvertens** [comment]
I have an update on action 2. 
- I am using a ctsm5.1dev130-noresm_v1 from https://github.com/NorESMhub/CTSM
- I tried to use a spun up CLM dataset that Rosie provided and the model crashed in FATES. I can provide that error if someone wants it.
- I decided to simply to a 10 year spin up starting from a cold start using the following test.
SMS.f45_f45_mg37.2000_DATM%GSWP3v1_CLM51%FATES-NOCOMP_SICE_SOCN_SROF_SGLC_SWAV_SESP.betzy_intel.clm-FatesColdNoCompNoFire
- Starting from the restart file at year 10 - I did 2 runs. A 4 time step initial run. And a 2 time step initial run following by a 2 timestep restart run. I was able to show that both the restart files at timestep 3 and the history files at timestep 3 (clm2.h0.2000-01-01-05400.nc)  were different in a small number of variables. 
- The history files comparison showed the following diffs: 
 ```
RMS FATES_CBALANCE_ERROR             2.6008E-18            NORMALIZED  2.3165E+00
 RMS FATES_DISTURBANCE_RATE_P2P       3.3688E-03            NORMALIZED  2.0000E+00
 RMS FATES_DISTURBANCE_RATE_POTENTIAL 3.3688E-03            NORMALIZED  2.0000E+00
 RMS FATES_DISTURBANCE_RATE_TREEFALL  3.3688E-03            NORMALIZED  2.0000E+00
 RMS FATES_ERROR_EL                   2.6008E-18            NORMALIZED  2.3165E+00
 RMS FATES_FRAGMENTATION_SCALER_SL    2.4246E-01            NORMALIZED  2.0358E+00
 RMS FATES_PRIMARY_PATCHFUSION_ERR    8.1046E-14            NORMALIZED  2.0000E+00
```
- The restart files showed the following diffs:
 ```
RMS tempavg_bgnpp                           NaN            NORMALIZED  0.0000E+00
 RMS annsum_counter_ch4                      NaN            NORMALIZED  0.0000E+00
 RMS tempavg_somhr                           NaN            NORMALIZED  0.0000E+00
 RMS tempavg_finrw                           NaN            NORMALIZED  0.0000E+00
 RMS fates_fcansno_pa                 1.6970E-03            NORMALIZED  1.7092E-07
 RMS fates_solar_zenith_angle_pa      1.5994E-04            NORMALIZED  1.6109E-08
 RMS fates_gnd_alb_dif                6.2440E-03            NORMALIZED  1.5925E+00
 RMS fates_gnd_alb_dir                6.3061E-03            NORMALIZED  1.6566E+00
 RMS fates_errfates_vec_001           9.5475E-14            NORMALIZED  5.6401E+00
```

I am really perplexed as to why the albedos are different. I really would need some advice as to how to proceed with this. I have not looked at FATES for several years. I am happy to add new variables to the restart files. Would another chat help?
Should we wait for Rosie to come back?

**Comment by mvdebolskiy** [comment]
@rgknox can this be due to #428 and ESCOMP/CTSM#548 ? Do you remember what were the tests performed when merging that PR?

**Comment by ckoven** [comment]
@mvertens one question I have is on the time stepping that you did. Much of the FATES-specific processes only occur once per day, so every 48th timestep. Was the daily timestep one of the two timesteps in your restart run? Because if not, that says that the daily FATES code shouldn't even be triggered in the experiment you did, despite you showing differences in things like the history-reported disturbance rates? So if that is the case it could considerably narrow the set off possible errors here.

**Comment by mvertens** [comment]
@ckoven - the daily time step was not one of the two timesteps in my restart. I started at the beginning of the day and did 4 timesteps in one run and 2/2 in the restart run. I then compared the two timesteps at timestep 4.

**Comment by rgknox** [comment]
@mvdebolskiy that could be a good lead.  We had a test suit in 2018 similar to what we have now, but it may had not contained tests long enough.

**Comment by rgknox** [comment]
@mvdebolskiy , I'm noticing that there is some similar code doing the same thing in the call to PatchNorman radiation during the restart and the normal call sequence.  Even if this is not the problem we are tackling in this thread, I think it would be better to have the zero'ing and initializations contained in one routine that is called from both locations.

See:  https://github.com/NGEET/fates/blob/main/main/FatesRestartInterfaceMod.F90#L3588-L3644
See: https://github.com/NGEET/fates/blob/main/biogeophys/EDSurfaceAlbedoMod.F90#L106-L164

**Comment by rgknox** [comment]
I ran a test where I modify the line linked in my previous post to match that the call in EDSurfaceAlbedoMod, but it did not remove the error. I think we are getting closer though, the diffs that you uncovered in the restart @mvertens do seem to point to something in the radiation restart.

**Comment by mvdebolskiy** [comment]
@rgknox 
In the `ED_Norman_Radiation()` assignment starts from 1st pft.
https://github.com/NGEET/fates/blob/21e18c676b5e87999faead5fad9c610df4fbf255/biogeophys/EDSurfaceAlbedoMod.F90#L100

In the ` update_3dpatch_radiation()`  there is no such condition. Do I understand correctly, that when reading from the restart bareground pft will get the values from ` update_3dpatch_radiation()` in nocomp mode while in competition mode it will get just clm restart values?

**Comment by rosiealice** [comment]
A thought. If this is a radiation code thing, does it fail the test with the new two stream scheme?

**Comment by rgknox** [comment]
A couple of other things I tried:
Running with only evergreens: failure persists
*Running in debug mode with gnu: failure persists

I ran the latter test because I was concerned the radiation scheme may had been pulling from  indices outside of the allocation bounds on arrays, and thus mixing in un-initialized data. It was a long shot, but this was not the issue.

**Comment by ckoven** [comment]
@mvertens this is probably a separate issue from the fundamental one(s) responsible for the problems here, but the history fields `FATES_DISTURBANCE_RATE_P2P`, `FATES_DISTURBANCE_RATE_POTENTIAL`, `FATES_DISTURBANCE_RATE_TREEFALL` that are showing up as non-bit-for-bit in your experiment despite being tied to the daily timestep are all either removed or reworked as part of #1040. I will put in code to make sure that the underlying variable for these is restarted correctly into that PR.

**Comment by rgknox** [comment]
That could certainly explain the diffs. I see we call restart() in the clmfates_interfaceMod, and in that routine we call update_site(), which calls termination here:
https://github.com/NGEET/fates/blob/sci.1.67.5_api.27.0.0/main/EDMainMod.F90#L808-L809

Looking for more instances.

**Comment by mvdebolskiy** [comment]
@rgknox More specifically.  The termination happens in [canopy_structure()](https://github.com/NGEET/fates/blob/a3048a6469e868144878bfc6ca497c89fa975f3f/main/EDMainMod.F90#L793C13-L793C30), in the [call index 14](https://github.com/NGEET/fates/blob/a3048a6469e868144878bfc6ca497c89fa975f3f/biogeochem/EDCanopyStructureMod.F90#L245) after the cohort fuse.

**Comment by rgknox** [comment]
@ckoven and I were talking about the call to canopy_structure(), and in theory, we shouldn't need that call during the restart read. There is no need to perform promotions/demotions and such during a restart, nor terminations.  The information on canopy position and such could be retained in the restart file.   
I am curious though, why it would be terminating here, one would think that the cohort being terminated would had been already terminated prior to the restart write.

**Comment by mvdebolskiy** [comment]
@rgknox  
I've tried to change [the condition in the clm_driver](https://github.com/ESCOMP/CTSM/blob/03e9fd54cea00687418a791972e402a97967f62f/src/main/clm_driver.F90#L1135) to is_end_curr_day() but it did not help.

**Comment by mvdebolskiy** [comment]
Also, I am a bit confused about why the gridcell index is used [here](https://github.com/ESCOMP/CTSM/blob/03e9fd54cea00687418a791972e402a97967f62f/src/utils/clmfates_interfaceMod.F90#L1592). I understand that right now there is only one site per column per gridcell and it's the first one among the columns.

**Comment by rgknox** [comment]
@mvdebolskiy , that is because we currently only have 1 naturally vegetated column on each grid cell, thus only 1 fates column per grid cell. Thus we allocate enough restart memory for all the patches and cohorts, assuming we only need one set of them on each grid-cell.

**Comment by rgknox** [comment]
Note, I tried a test using two-stream instead of norman radiation, does not have an effect on the restart compare.

**Comment by rosiealice** [comment]
It is indeed confusing why things should be terminating on restart and why this is triggered after and not before the restart occurs. I also agree that we should not in principle need to call canopy_structure during the restarting process. Doing away with it might prevent this termination occurring, but might actually mask the reason why the conditions wrt the termination change? Probably would be useful to track down what is triggering the termination here?

**Comment by mvdebolskiy** [comment]
[fates branch I am working in](https://github.com/mvdebolskiy/fates/tree/fix_restart_bug)
[ctsm](https://github.com/mvdebolskiy/CTSM/tree/fate_rest_bug_fix)

**Comment by mvdebolskiy** [comment]
Things from today's meeting
- [ ] Chek if init_cold is not called during restart read
- [ ] Make `NCL_p` a restart var and switch of `canopy_structure` call during restart read
- [ ] try to fuse on the restart write to see if there are cohorts to fuse
- [ ] dump small `n` cohorts into log on restart write and read
- [ ] Get rid of the variables `FATES_DISTURBANCE_RATE_P2P, FATES_DISTURBANCE_RATE_POTENTIAL, FATES_DISTURBANCE_RATE_TREEFALL` from history files, so the test don't fail because of them

**Comment by glemieux** [comment]
During the meeting I ran a 110 day `1x1_brazil` at @rosiealice suggestion to see if we can narrow down to a specific site to make troubleshooting a little easier.   I wasn't able to get the issue to replicate in this configuration.

I also tried adding a check for `hlm_is_restart` in `canopy_structure` to avoid the `do while(area_not_balanced)` loop, hoping to avoid the termination calls here:
https://github.com/NGEET/fates/blob/a3048a6469e868144878bfc6ca497c89fa975f3f/biogeochem/EDCanopyStructureMod.F90#L190-L300

This hit an `endrun` during the restart here almost immediately after restart initialization:
https://github.com/NGEET/fates/blob/a3048a6469e868144878bfc6ca497c89fa975f3f/biogeochem/EDCanopyStructureMod.F90#L1421-L1427

cheyenne folder: `/glade/u/home/glemieux/scratch/ctsm-tests/tests_issue1051_d110_f45_areabalavoid`

**Comment by rgknox** [comment]
@glemieux , are you saying that the model crashes if we don't call that promotion/demotion code during the restart sequence?

**Comment by rgknox** [comment]
my tests are now passing when bypassing canopy_structure() via update_site() on the restart, as well as adding ncl_p to the restart

**Comment by rgknox** [comment]
Can someone else test these branches to see if these fixes work for them? I'll try some more tests, longer tests to see if that works:

https://github.com/rgknox/ctsm/tree/fates-nocomp-fix

The fates branch was added to the externals of the clm branch btw.

https://github.com/rgknox/fates/tree/nocomp-fix-tests

**Comment by glemieux** [comment]
> @glemieux , are you saying that the model crashes if we don't call that promotion/demotion code during the restart sequence?

@rgknox yes that's what I was seeing.  This was with the latest ctsm tag using sci.1.67.2_api.27.0.0.

**Comment by glemieux** [comment]
@rgknox I confirmed that the branch fix worked for me as well.  This was a 110, f45 test on Cheyenne.
Cheyenne folder: `/glade/u/home/glemieux/scratch/ctsm-tests/tests_issue1051_d110_f45_rgknox-nocompfix`

**Comment by rosiealice** [comment]
Amazing!

**Comment by mvdebolskiy** [comment]
Checked that `init_cold` is not called on restart. Will try to test rgknox's fix with longer test over the weekend.

**Comment by mvertens** [comment]
Fantastic news!!!!

lør. 14. okt. 2023 kl. 15:06 skrev mvdebolskiy ***@***.***>:

**Comment by mvdebolskiy** [comment]
Quick question: 
Do I understand correctly that canopy layers `1 ... nclmax` are going in the order from tallest to shortest?

**Comment by rgknox** [comment]
yes, canopy 1 is the upper-most canopy layer, and then it works downwards.

**Comment by rgknox** [comment]
While #1098 seems to enable some long restarts to pass, it does not seem to address the whole problem.  
I do generate fails for longer ERS tests, for example 11 month runs with full fates on an f45. 
I'm planning a few different tests:
1) trying with fates_comp_excln -1 (ie strict rank based promotion/demotion) instead of probabilistic
2) again, removing deciduous for evergreen only
3) I may try with fixed biogeog/nocomp, (my full fates runs generate the restart fail), but I believe @mvdebolskiy might be generating restart compare fails with this setting already?

**Comment by rgknox** [comment]
I believe that this call to canopy spread is also a problem for restarts: https://github.com/NGEET/fates/blob/main/main/EDMainMod.F90#L789

If you look at the spread calculation, it is an "incrementer", ie it will modify the spread a little more until the canopy goes all the way towards closed. This is problematic if it is called more times in the restart sequence than the original sequence. I turned this call off on the restart, and the ERS 11 month test passed. I'll add this change to the testing branch and continue to see if we can find more problems.

**Comment by rosiealice** [comment]
Hi Ryan, 

Nice catch, this sounds like a very viable culprit to me...

**Comment by mvertens** [comment]
Hi Ryan,

Great work! Do we want to do a longer run to verify that restarts are good
out to longer times?

Mariana

man. 30. okt. 2023 kl. 09:10 skrev Rosie Fisher ***@***.***>:

**Comment by rgknox** [comment]
@mvertens I think that's a great idea.  I was wondering how the queue wall-time is calculated for the tests, to make sure that we don't push up against the maximum. And I was also playing around with maximizing run time while keeping the wall-time somewhat consistent with other tests.  For instance in that Lm25 I tried, I doubled the core count by using the P144x1 setting, and that took around 1000 seconds  (17 minutes), not too bad. I think that is a reasonable wall-time to include in the test suite.  Any ideas/brainstorming are welcome.
I was also going to try creating a test that initializes the vegetation demographics from initialization files. I think this will go a long way towards making our tests more robust.  If we do this, I'd like us to integrate the pull request that updates allometry (#1093) before we set  the initialization files up.
